### ISSUE TITLE
Not getting any trace from MongoDB connection in a NestJS application, Instrumentation is not working

### ISSUE THREAD

**Issue by divyanshunegi** [open-telemetry/opentelemetry-js-contrib#1469]
I have integrated Opentelemetry Instrumentation for MongoDB, using `MongoDBInstrumentation` as well from `MongooseInstrumentation` but none of the instrumentation is working for me and I am unable to see any traces for DB for my service, whereas the `postgresql` traces are working fine on the same service without adding any additional instrumentation.

### What version of OpenTelemetry are you using?
```
  "@opentelemetry/api": "^1.4.0",
    "@opentelemetry/auto-instrumentations-node": "^0.36.3",
    "@opentelemetry/exporter-metrics-otlp-http": "^0.37.0",
    "@opentelemetry/exporter-trace-otlp-http": "^0.35.1",
    "@opentelemetry/host-metrics": "^0.31.0",
    "@opentelemetry/resources": "^1.9.1",
    "@opentelemetry/sdk-metrics": "^1.11.0",
    "@opentelemetry/sdk-node": "^0.35.1",
    "@opentelemetry/semantic-conventions": "^1.9.1",
    "@opentelemetry/tracing": "^0.24.0",
```

### What version of Node are you using?
❯ node -v
`v16.15.0`

### What did you do?

below is the singleton class to initialise the Telemetry, where I am using this code to setup the instrumentation

```
    this.sdk = new opentelemetry.NodeSDK({
      traceExporter,
      instrumentations: [
        getNodeAutoInstrumentations(),
        new MongoDBInstrumentation(),
        // new MongooseInstrumentation(),
      ],
      resource: resource,
    });
```

But no matter what I use `MongoDBInstrumentation` vs `MongooseInstrumentation` none of the instrumentation is passing data to our dashboard. but the data from `postgresql` traces is visible clearly.

```ts
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { HostMetrics } from '@opentelemetry/host-metrics';
import {
  ConsoleMetricExporter,
  MeterProvider,
  PeriodicExportingMetricReader,
} from '@opentelemetry/sdk-metrics';
import { Resource } from '@opentelemetry/resources';
import * as opentelemetry from '@opentelemetry/sdk-node';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import config from 'config';
import { errorLog, infoLog } from 'src/shared/logger/logger.helpers';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { Meter, ObservableGauge, ObservableResult } from '@opentelemetry/api';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { MongooseInstrumentation } from '@opentelemetry/instrumentation-mongoose';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { MongoDBInstrumentation } from '@opentelemetry/instrumentation-mongodb';

enum ECustomerMeter {
  node_process = 'node.process',
}

enum ECustomMetric {
  process_mem_rss = 'process.memory.rss',
}

interface IHostMetricsConfig {
  collectorOptions: {
    url: string;
    concurrencyLimit: number;
  };
  metricReader: {
    intervalInMs: number;
  };
  useConsoleMetricExporter: boolean;
}
export class APMSignoz {
  private static instance: APMSignoz;
  private readonly sdk: opentelemetry.NodeSDK;
  private readonly signoz = config.get<{ baseUrl: string; enabled: boolean }>(
    'signoz',
  );
  private readonly hostMetricConfig =
    config.get<IHostMetricsConfig>('hostMetrics');
  private readonly hostMetrics: HostMetrics | undefined;
  private readonly metricMeterProvider: MeterProvider | undefined;
  private readonly customProcessMeter: Meter | undefined;
  private readonly customProcessMemoryUsageRssGauge:
    | ObservableGauge
    | undefined;

  private constructor() {
    this.observeRss = this.observeRss.bind(this);
    // Configure the SDK to export telemetry data to the console
    // Enable all auto-instrumentations from the meta package
    const exporterOptions = {
      url: this.signoz.baseUrl,
    };

    const svcName = `${config.get<string>('service.name')}:${
      process.env.NODE_ENV
    }`;

    const traceExporter = new OTLPTraceExporter(exporterOptions);
    const resource = new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: svcName,
    });

    this.sdk = new opentelemetry.NodeSDK({
      traceExporter,
      instrumentations: [
        getNodeAutoInstrumentations(),
        new MongoDBInstrumentation(),
        // new MongooseInstrumentation(),
      ],
      resource: resource,
    });

    // host metrics
    if (this.signoz.enabled) {
      const metricExporter = !this.hostMetricConfig.useConsoleMetricExporter
        ? new OTLPMetricExporter(this.hostMetricConfig.collectorOptions)
        : new ConsoleMetricExporter();
      this.metricMeterProvider = new MeterProvider({
        resource,
      });
      this.metricMeterProvider.addMetricReader(
        new PeriodicExportingMetricReader({
          exporter: metricExporter,
          exportIntervalMillis: this.hostMetricConfig.metricReader.intervalInMs,
        }),
      );
      this.hostMetrics = new HostMetrics({
        meterProvider: this.metricMeterProvider,
        name: `${svcName}-host-metrics`,
      });

      // documentation for custom metrics: https://opentelemetry.io/docs/instrumentation/js/instrumentation/#using-observable-async-gauges
      this.customProcessMeter = this.metricMeterProvider?.getMeter(
        ECustomerMeter.node_process,
      );
      this.customProcessMemoryUsageRssGauge =
        this.customProcessMeter?.createObservableGauge(
          ECustomMetric.process_mem_rss,
        );
    }
  }

  public static getInstance(): APMSignoz {
    if (!APMSignoz.instance) {
      APMSignoz.instance = new APMSignoz();
    }

    return APMSignoz.instance;
  }

  async start() {
    if (this.signoz.enabled) {
      this.customProcessMemoryUsageRssGauge?.addCallback(this.observeRss);
      this.hostMetrics?.start();
      await this.sdk
        .start()
        .then(() => infoLog('Tracing initialized'))
        .catch((error) => errorLog(error, 'Error initializing tracing'));
    }
  }

  private observeRss(observableResult: ObservableResult) {
    const rss = process.memoryUsage().rss;
    observableResult.observe(rss);
  }

  async stop() {
    if (!this.signoz.enabled) {
      return;
    }
    this.customProcessMemoryUsageRssGauge?.removeCallback(this.observeRss);
    this.metricMeterProvider?.shutdown();
    await this.sdk
      .shutdown()
      .then(() => infoLog('Tracing terminated'))
      .catch((error) => errorLog(error, 'Error terminating tracing'));
  }
}

```

### What did you expect to see?

I need to see the mongoDB call traces without this data the dashboard for us is not very useful, in case we have an API call which is taking longer time we want to see the traces on which call took the max time with mongoDB  traces its possible.

### What did you see instead?

Nothing, No trace at all from mongoDB

### Additional context

UPDATE

In our service we are using the following dependencies 

```
    "mongoose": "^6.2.9",
    "@nestjs/mongoose": "^9.0.3",
    "migrate-mongo": "^9.0.0",
    "mongodb-memory-server": "^8.8.0",
```

**Comment by pichlermarc** [comment]
Hi, thanks for reaching out.

> Not sure what version of Mongoose you are using But see - [Issue around Mongoose 7](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1420)

I would echo this, having information would definitely help us reproduce the issue :slightly_smiling_face:

**Comment by divyanshunegi** [comment]
> I can confirm that I had to roll back to Mongoose6 to get span/traces reporting.... Mongoose 7...no go. The only change was the version of Mongoose- Nothing else changed

```
   "mongoose": "^6.2.9",
    "@nestjs/mongoose": "^9.0.3",
    "migrate-mongo": "^9.0.0",
    "mongodb-memory-server": "^8.8.0",
```
using the following dependencies currently, mongoose is already `6.2.9`

**Comment by divyanshunegi** [comment]
> Hi, thanks for reaching out.
> 
> > Not sure what version of Mongoose you are using But see - [Issue around Mongoose 7](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1420)
> 
> I would echo this, having information would definitely help us reproduce the issue 🙂

Updated the description, 
```
 "mongoose": "^6.2.9",
    "@nestjs/mongoose": "^9.0.3",
    "migrate-mongo": "^9.0.0",
    "mongodb-memory-server": "^8.8.0",
```

**Comment by pichlermarc** [comment]
@blumamir would you mind looking into this since you're listed as the component owner? :thinking:

**Comment by blumamir** [comment]
@divyanshunegi  [This blog post](https://opentelemetry.io/blog/2022/troubleshooting-nodejs/) has various methods on how to troubleshot such issues.

I would recommend that first you enable the diag logging to verify if these instrumentations are indeed loaded. For mongoose, you should be able to see [these diag records](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/8d9687d89e4a80dbf2a5e8be6fb027ff20824593/plugins/node/instrumentation-mongoose/src/mongoose.ts#L147).

Above you attached the code to initialize opentelemetry, but how and when do you execute it? 
is it possible that your mongodb / mongoose are "require"d by nestjs prior to you registering the instrumentation libraries? 
If I remember correctly, the instrumentations are loaded only when you call `sdk.start()` and this MUST happen BEFORE any import of the instrumented libraries. Can you please add prints to verify the order is indeed correct in this case?

**Comment by bradleybeighton** [comment]
I'm also having an issue. I've confirmed that the sdk.start() happens before any import of modules.

**Comment by bradleybeighton** [comment]
@blumamir @pichlermarc  

After some trial and error, I found that changing my import from 
`
import { initTracingSDK } from '...';
`
to 
`
import * as tracer from '....';
tracer.initTracingSDK();
`

Seemed to resolve the issues.

**Comment by dyladan** [comment]
This issue hasn't gotten any activity so I'm going to assume the above workaround worked. If it doesn't please feel free to comment here and we'll reopen this or you can create a new issue linking to this one.

**Comment by daniel90180** [comment]
> [@blumamir](https://github.com/blumamir) [@pichlermarc](https://github.com/pichlermarc)
> 
> After some trial and error, I found that changing my import from `import { initTracingSDK } from '...';` to `import * as tracer from '....'; tracer.initTracingSDK();`
> 
> Seemed to resolve the issues.

Which package/import path did you do this for?
